wolox-books/src/app/screens/sign-up/sign-up.component.spec.ts

```diff
@@ -1,25 +1,202 @@
-
+import { MockUserService } from './mocks/mock-user-service';
+import { User } from './interfaces/user';
+import { MockUser } from './mocks/mock-user';
+import { RouterTestingModule } from '@angular/router/testing';
+import { UserService } from './../../services/user.service';
+import { SignBaseModule } from './../../components/sign-base/sign-base.module';
+import { FormsModule, ReactiveFormsModule } from '@angular/forms';
 import { SignUpComponent } from './sign-up.component';
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { HttpClientTestingModule } from '@angular/common/http/testing';
 
 describe('SignUpComponent', () => {
-  let component: SignUpComponent;
   let fixture: ComponentFixture<SignUpComponent>;
+  let component: SignUpComponent;
+
+  function fillForm(mockUser: User): void {
+    component.form.controls.first_name.setValue(mockUser.first_name);
+    component.form.controls.last_name.setValue(mockUser.last_name);
+    component.form.controls.email.setValue(mockUser.email);
+    component.form.controls.password.setValue(mockUser.password);
+    component.form.controls.password_confirmation.setValue(mockUser.password_confirmation);
+  }
 
-  beforeEach(async () => {
-    await TestBed.configureTestingModule({
-      declarations: [ SignUpComponent ]
-    })
-    .compileComponents();
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      declarations: [
+        SignUpComponent
+      ],
+      imports: [
+        FormsModule,
+        ReactiveFormsModule,
+        HttpClientTestingModule,
+        SignBaseModule,
+        RouterTestingModule
+      ],
+      providers: [
+        { provide: UserService, useClass: MockUserService}
+      ]
+    }).compileComponents();
   });
 
   beforeEach(() => {
     fixture = TestBed.createComponent(SignUpComponent);
     component = fixture.componentInstance;
-    fixture.detectChanges();
   });
 
-  it('should create', () => {
-    expect(component).toBeTruthy();
+  describe('Test: when the component is started', () => {
+    it('should create the SignUpComponent', () => {
+      expect(component).toBeTruthy();
+    });
+
+    it('A form should be with first name, last name, email, password, password confirmation input and sign up button', () => {
+      const firstName = fixture.debugElement.nativeElement.querySelector('.form__input[formControlName=first_name]');
+      const lastName = fixture.debugElement.nativeElement.querySelector('.form__input[formControlName=last_name]');
+      const email = fixture.debugElement.nativeElement.querySelector('.form__input[formControlName=email]');
+      const password = fixture.debugElement.nativeElement.querySelector('.form__input[formControlName=password]');
+      const passwordConfirmation = fixture.debugElement.nativeElement.querySelector('.form__input[formControlName=password_confirmation]');
+      const signUpButton = fixture.debugElement.nativeElement.querySelector('.form__submit');
```

Comment:
A very long sentence is repeated a lot. `fixture.debugElement.nativeElement.querySelector` is present 6 times.
Please consider coding this more dynamic.
```js
const selectors = [
  '.form__input[formControlName=first_name]',
   '.form__input[formControlName=last_name]',
  ...
]
const [firstName, lastName, ...] = selectors.map(
  selector => fixture.debugElement.nativeElement.querySelector(selector)
);
```

Reply:
I liked your solution and based on it I decided to do something similar.
```js
elements = selectors.reduce((a, c) => {
    a[c.name] = fixture.debugElement.nativeElement.querySelector(c.className);
    return a;
}, {});
...
expect(elements.firstName).toBeDefined();
```